test_util/test_azure.py

```diff
@@ -0,0 +1,139 @@
+    options.ssh_key_path = os.getenv('DCOS_SSH_KEY_PATH', 'ssh_key')
+    options.location = os.getenv('AZURE_LOCATION', 'East US')
+    options.linux_user = os.getenv('AZURE_LINUX_USER', 'dcos')
+    # Prefixes must not begin with a number
+    options.agent_prefix = os.getenv('AZURE_AGENT_PREFIX', 'test' + random_id(10).lower())
+    options.master_prefix = os.getenv('AZURE_MASTER_PREFIX', 'test' + random_id(10).lower())
+    options.vm_size = os.getenv('AZURE_VM_SIZE', 'Standard_D2')
+    options.num_agents = int(os.getenv('AGENTS', '2'))
+    options.name_suffix = os.getenv('AZURE_DCOS_SUFFIX', '12345')
+    options.oauth_enabled = os.getenv('AZURE_OAUTH_ENABLED', 'false') == 'true'
+    options.vm_diagnostics_enabled = os.getenv('AZURE_VM_DIAGNOSTICS_ENABLED', 'true') == 'true'
+    options.azure_cleanup = os.getenv('AZURE_CLEANUP', 'true') == 'true'
+    options.ci_flags = os.getenv('CI_FLAGS', '')
+
+    add_env = {}
+    prefix = 'TEST_ADD_ENV_'
+    for k, v in os.environ.items():
+        if k.startswith(prefix):
+            add_env[k.replace(prefix, '')] = v
+    options.add_env = add_env
+    options.pytest_cmd = os.getenv('DCOS_PYTEST_CMD', "py.test -rs -vv -m 'not ccm' ") + os.getenv('CI_FLAGS', '')
+    return options
+
+
+def main():
+    options = check_environment()
+    aw = AzureWrapper(
+        options.subscription_id,
+        options.tenant_id,
+        options.client_id,
+        options.client_secret)
+    arm_deploy = TemplateDeployment.create(
+        azure_wrapper=aw,
+        template_uri=options.template_url,
+        location=options.location,
+        name=options.name,
+        public_key=options.public_ssh_key,
+        master_prefix=options.master_prefix,
+        agent_prefix=options.agent_prefix,
+        admin_name=options.linux_user,
+        oauth_enabled=options.oauth_enabled,
+        vm_size=options.vm_size,
+        agent_count=options.num_agents,
+        name_suffix=options.name_suffix,
+        vm_diagnostics_enabled=options.vm_diagnostics_enabled,
+        rollback=options.azure_cleanup)
+    result = 1
+    try:
+        with SSHTunnel(options.linux_user, options.ssh_key_path, arm_deploy.outputs['masterFQDN'], port=2200) as t:
+            result = integration_test(
+                tunnel=t,
+                test_dir='/home/{}'.format(options.linux_user),
+                dcos_dns=arm_deploy.get_master_ips()[0],
+                master_list=arm_deploy.get_master_ips(),
+                agent_list=arm_deploy.get_private_ips(),
+                public_agent_list=arm_deploy.get_public_ips(),
+                provider='azure',
+                test_dns_search=False,
+                add_env=options.add_env,
+                pytest_cmd=options.pytest_cmd)
+    finally:
+        if result == 0:
+            log.info('Test successsful! Deleting CloudFormation...')
+        else:
+            logging.warning('Test exited with an error')
+        if options.azure_cleanup or result == 0:
```

Comment:
Can we maybe use https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack rather than having to write the big try / finally with complicated ending?